src/chat_interface/chatgpt.ts

```diff
@@ -0,0 +1,87 @@
+import ChatInterface from "./chat_interface";
+import Role from "../enums/role";
+import type OpenAI from "openai";
+import type RateLimiter from "../rate_limiter";
+
+export default class ChatGPT extends ChatInterface {
+    model: OpenAI;
+
+    chatParams: OpenAI.ChatCompletionCreateParamsNonStreaming | null;
+
+    history: OpenAI.ChatCompletionMessageParam[];
+
+    rateLimiter: RateLimiter;
+
+    constructor(model: OpenAI, rateLimiter: RateLimiter) {
+        super();
+        this.model = model;
+        this.chatParams = null;
+        this.history = [];
+        this.rateLimiter = rateLimiter;
+    }
+
+    startChat(params: OpenAI.ChatCompletionCreateParamsNonStreaming): void {
+        this.chatParams = params;
+        if (params.messages.length > 0) {
+            this.history = params.messages;
+        }
+    }
+
+    async sendMessage(message: string): Promise<string> {
+        if (!this.chatParams) {
+            console.trace("Chat not started");
+            return "";
+        }
+
+        await this.rateLimiter.wait();
+        this.rateLimiter.apiCalled();
+        this.history.push({ role: Role.User, content: message });
+
+        try {
+            const response = await this.model.chat.completions.create({
+                ...this.chatParams,
+                messages: this.history,
+            });
+
+            const responseText = response.choices[0].message.content;
+            if (!responseText) {
+                return "";
+            }
+
+            this.history.push({ role: Role.System, content: responseText });
+            return responseText;
+        } catch (err) {
+            console.error(err);
```

Comment:
nit: itll be easier to debug with more context on where this error is being thrown from. We might be lucky and the error might imply it, but not guaranteed